lucet-wasi/src/hostcalls.rs

```diff
@@ -980,6 +985,211 @@ pub extern "C" fn __wasi_random_get(
     return wasm32::__WASI_ESUCCESS;
 }
 
+nix::ioctl_read_bad!(fionread, nix::libc::FIONREAD, c_int);
+
+fn wasi_clock_to_relative_ns_delay(
+    wasi_clock: host::__wasi_subscription_t___wasi_subscription_u___wasi_subscription_u_clock_t,
+) -> u128 {
+    if wasi_clock.flags != wasm32::__WASI_SUBSCRIPTION_CLOCK_ABSTIME {
+        return wasi_clock.timeout as u128;
+    }
+    let now: u128 = SystemTime::now()
+        .duration_since(SystemTime::UNIX_EPOCH)
+        .expect("Current date is before the epoch")
+        .as_nanos();
+    let deadline = wasi_clock.timeout as u128;
+    deadline.saturating_sub(now)
+}
+
+#[no_mangle]
+pub extern "C" fn __wasi_poll_oneoff(
+    vmctx: *mut lucet_vmctx,
+    input: wasm32::uintptr_t,
+    output: wasm32::uintptr_t,
+    nsubscriptions: wasm32::size_t,
+    nevents: wasm32::uintptr_t,
+) -> wasm32::__wasi_errno_t {
+    if nsubscriptions as u64 > wasm32::__wasi_filesize_t::max_value() {
+        return wasm32::__WASI_EINVAL;
+    }
+    let mut vmctx = unsafe { Vmctx::from_raw(vmctx) };
+    unsafe { enc_pointee(&mut vmctx, nevents, 0) }.unwrap();
+    let input_slice_ =
+        unsafe { dec_slice_of::<wasm32::__wasi_subscription_t>(&mut vmctx, input, nsubscriptions) }
+            .unwrap();
+    let input_slice = unsafe { slice::from_raw_parts(input_slice_.0, input_slice_.1) };
+
+    let output_slice_ =
+        unsafe { dec_slice_of::<wasm32::__wasi_event_t>(&mut vmctx, output, nsubscriptions) }
+            .unwrap();
+    let output_slice = unsafe { slice::from_raw_parts_mut(output_slice_.0, output_slice_.1) };
+
+    let input: Vec<_> = input_slice.iter().map(|x| dec_subscription(x)).collect();
+
+    let timeout = input
+        .iter()
+        .filter(|event| match event {
+            Ok(event) => event.type_ == wasm32::__WASI_EVENTTYPE_CLOCK,
+            _ => false,
+        })
+        .map(|event| {
+            let event = event.unwrap();
+            unsafe {
+                (
+                    wasi_clock_to_relative_ns_delay(event.u.clock) / 1_000_000,
+                    event.userdata,
```

Comment:
`unsafe` should go here, rather than around the whole tuple

Reply:
done, thanks!